File: API/controllers/postsController.js
const { query } = require('../config/database');

// Obtener todos los posts publicados
async function getPublishedPosts(req, res) {
  try {
    const { page = 1, limit = 10, category } = req.query;
    const offset = (page - 1) * limit;
    
    let sql = `
      SELECT 
        p.id, p.title, p.slug, p.excerpt, p.featured_image, 
        p.view_count, p.created_at, p.published_at,
        c.name as category_name, c.slug as category_slug, c.color as category_color,
        u.full_name as author_name
      FROM posts p
      LEFT JOIN categories c ON p.category_id = c.id
      LEFT JOIN users u ON p.author_id = u.id
      WHERE p.status = 'published'
    `;
    
    const params = [];
    
    if (category) {
      sql += ` AND c.slug = $${params.length + 1}`;
      params.push(category);
    }
    
    sql += ` ORDER BY p.published_at DESC LIMIT $${params.length + 1} OFFSET $${params.length + 2}`;
    params.push(parseInt(limit), offset);
    
    const result = await query(sql, params);
    
    // Obtener el total de posts para paginación
    let countSql = `
      SELECT COUNT(*) as total
      FROM posts p
      LEFT JOIN categories c ON p.category_id = c.id
      WHERE p.status = 'published'
    `;
    
    if (category) {
      countSql += ` AND c.slug = $1`;
    }
    
    const countResult = await query(countSql, category ? [category] : []);
    const total = parseInt(countResult.rows[0].total);
    
    res.json({
      success: true,
      data: result.rows,
      pagination: {
        page: parseInt(page),
        limit: parseInt(limit),
        total,
        pages: Math.ceil(total / limit)
      }
    });
    
  } catch (error) {
    console.error('Error obteniendo posts:', error);
    res.status(500).json({
      success: false,
      message: 'Error interno del servidor'
    });
  }
}

// Obtener un post específico por slug
async function getPostBySlug(req, res) {
  try {
    const { slug } = req.params;
    
    // Incrementar el contador de vistas
    await query(
      'UPDATE posts SET view_count = view_count + 1 WHERE slug = $1',
      [slug]
    );
    
    const sql = `
      SELECT 
        p.id, p.title, p.slug, p.content, p.excerpt, p.featured_image,
        p.view_count, p.created_at, p.published_at,
        c.name as category_name, c.slug as category_slug, c.color as category_color,
        u.full_name as author_name, u.bio as author_bio
      FROM posts p
      LEFT JOIN categories c ON p.category_id = c.id
      LEFT JOIN users u ON p.author_id = u.id
      WHERE p.slug = $1 AND p.status = 'published'
    `;
    
    const result = await query(sql, [slug]);
    
    if (result.rows.length === 0) {
      return res.status(404).json({
        success: false,
        message: 'Post no encontrado'
      });
    }
    
    // Obtener comentarios aprobados
    const commentsSql = `
      SELECT id, content, author_name, created_at
      FROM comments
      WHERE post_id = $1 AND is_approved = true
      ORDER BY created_at DESC
    `;
    
    const commentsResult = await query(commentsSql, [result.rows[0].id]);
    
    const post = result.rows[0];
    post.comments = commentsResult.rows;
    
    res.json({
      success: true,
      data: post
    });
    
  } catch (error) {
    console.error('Error obteniendo post:', error);
    res.status(500).json({
      success: false,
      message: 'Error interno del servidor'
    });
  }
}

// Obtener posts relacionados (misma categoría)
async function getRelatedPosts(req, res) {
  try {
    const { slug } = req.params;
    const { limit = 3 } = req.query;
    
    const sql = `
      SELECT 
        p.id, p.title, p.slug, p.excerpt, p.created_at,
        c.name as category_name, c.slug as category_slug
      FROM posts p
      LEFT JOIN categories c ON p.category_id = c.id
      WHERE p.status = 'published' 
        AND p.slug != $1
        AND p.category_id = (
          SELECT category_id 
          FROM posts 
          WHERE slug = $1
        )
      ORDER BY p.published_at DESC
      LIMIT $2
    `;
    
    const result = await query(sql, [slug, parseInt(limit)]);
    
    res.json({
      success: true,
      data: result.rows
    });
    
  } catch (error) {
    console.error('Error obteniendo posts relacionados:', error);
    res.status(500).json({
      success: false,
      message: 'Error interno del servidor'
    });
  }
}

// Buscar posts
async function searchPosts(req, res) {
  try {
    const { q, page = 1, limit = 10 } = req.query;
    const offset = (page - 1) * limit;
    
    if (!q || q.trim().length < 2) {
      return res.status(400).json({
        success: false,
        message: 'El término de búsqueda debe tener al menos 2 caracteres'
      });
    }
    
    const searchTerm = `%${q.trim()}%`;
    
    const sql = `
      SELECT 
        p.id, p.title, p.slug, p.excerpt, p.created_at,
        c.name as category_name, c.slug as category_slug
      FROM posts p
      LEFT JOIN categories c ON p.category_id = c.id
      WHERE p.status = 'published'
        AND (p.title ILIKE $1 OR p.content ILIKE $1 OR p.excerpt ILIKE $1)
      ORDER BY p.published_at DESC
      LIMIT $2 OFFSET $3
    `;
    
    const result = await query(sql, [searchTerm, parseInt(limit), offset]);
    
    // Contar resultados totales
    const countSql = `
      SELECT COUNT(*) as total
      FROM posts p
      WHERE p.status = 'published'
        AND (p.title ILIKE $1 OR p.content ILIKE $1 OR p.excerpt ILIKE $1)
    `;
    
    const countResult = await query(countSql, [searchTerm]);
    const total = parseInt(countResult.rows[0].total);
    
    res.json({
      success: true,
      data: result.rows,
      pagination: {
        page: parseInt(page),
        limit: parseInt(limit),
        total,
        pages: Math.ceil(total / limit)
      }
    });
    
  } catch (error) {
    console.error('Error buscando posts:', error);
    res.status(500).json({
      success: false,
      message: 'Error interno del servidor'
    });
  }
}

module.exports = {
  getPublishedPosts,
  getPostBySlug,
  getRelatedPosts,
  searchPosts
}; 

File: DotnetDockerApi/Controllers/PostsController.cs
using System;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using DotnetDockerApi.Models;

namespace DotnetDockerApi.Controllers
{
    [ApiController]
    [Route("api/[controller]")]
    public class PostsController : ControllerBase
    {
        private readonly AppDbContext _context;

        public PostsController(AppDbContext context)
        {
            _context = context;
        }

        // GET: api/posts
        [HttpGet]
        public async Task<ActionResult<object>> GetPublishedPosts(
            [FromQuery] int page = 1, 
            [FromQuery] int limit = 10, 
            [FromQuery] string? category = null)
        {
            try
            {
                var query = _context.Posts
                    .Include(p => p.Category)
                    .Include(p => p.Author)
                    .Where(p => p.Status == "published");

                if (!string.IsNullOrEmpty(category))
                {
                    query = query.Where(p => p.Category!.Slug == category);
                }

                var total = await query.CountAsync();
                var posts = await query
                    .OrderByDescending(p => p.PublishedAt)
                    .Skip((page - 1) * limit)
                    .Take(limit)
                    .Select(p => new
                    {
                        p.Id,
                        p.Title,
                        p.Slug,
                        p.Excerpt,
                        p.FeaturedImage,
                        p.ViewCount,
                        p.CreatedAt,
                        p.PublishedAt,
                        CategoryName = p.Category!.Name,
                        CategorySlug = p.Category.Slug,
                        CategoryColor = p.Category.Color,
                        AuthorName = p.Author!.FullName
                    })
                    .ToListAsync();

                return Ok(new
                {
                    success = true,
                    data = posts,
                    pagination = new
                    {
                        page,
                        limit,
                        total,
                        pages = (int)Math.Ceiling((double)total / limit)
                    }
                });
            }
            catch (Exception ex)
            {
                return StatusCode(500, new
                {
                    success = false,
                    message = "Error interno del servidor"
                });
            }
        }

        // GET: api/posts/{slug}
        [HttpGet("{slug}")]
        public async Task<ActionResult<object>> GetPostBySlug(string slug)
        {
            try
            {
                var post = await _context.Posts
                    .Include(p => p.Category)
                    .Include(p => p.Author)
                    .Include(p => p.Comments.Where(c => c.IsApproved))
                    .FirstOrDefaultAsync(p => p.Slug == slug && p.Status == "published");

                if (post == null)
                {
                    return NotFound(new
                    {
                        success = false,
                        message = "Post no encontrado"
                    });
                }

                // Incrementar contador de vistas
                post.ViewCount++;
                await _context.SaveChangesAsync();

                var result = new
                {
                    post.Id,
                    post.Title,
                    post.Slug,
                    post.Content,
                    post.Excerpt,
                    post.FeaturedImage,
                    post.ViewCount,
                    post.CreatedAt,
                    post.PublishedAt,
                    CategoryName = post.Category!.Name,
                    CategorySlug = post.Category.Slug,
                    CategoryColor = post.Category.Color,
                    AuthorName = post.Author!.FullName,
                    AuthorBio = post.Author.Bio,
                    Comments = post.Comments.Select(c => new
                    {
                        c.Id,
                        c.Content,
                        c.AuthorName,
                        c.CreatedAt
                    })
                };

                return Ok(new
                {
                    success = true,
                    data = result
                });
            }
            catch (Exception ex)
            {
                return StatusCode(500, new
                {
                    success = false,
                    message = "Error interno del servidor"
                });
            }
        }

        // GET: api/posts/search
        [HttpGet("search")]
        public async Task<ActionResult<object>> SearchPosts(
            [FromQuery] string q, 
            [FromQuery] int page = 1, 
            [FromQuery] int limit = 10)
        {
            try
            {
                if (string.IsNullOrEmpty(q) || q.Trim().Length < 2)
                {
                    return BadRequest(new
                    {
                        success = false,
                        message = "El término de búsqueda debe tener al menos 2 caracteres"
                    });
                }

                var searchTerm = $"%{q.Trim()}%";

                var query = _context.Posts
                    .Include(p => p.Category)
                    .Where(p => p.Status == "published" &&
                               (EF.Functions.ILike(p.Title, searchTerm) ||
                                EF.Functions.ILike(p.Content, searchTerm) ||
                                EF.Functions.ILike(p.Excerpt!, searchTerm)));

                var total = await query.CountAsync();
                var posts = await query
                    .OrderByDescending(p => p.PublishedAt)
                    .Skip((page - 1) * limit)
                    .Take(limit)
                    .Select(p => new
                    {
                        p.Id,
                        p.Title,
                        p.Slug,
                        p.Excerpt,
                        p.CreatedAt,
                        CategoryName = p.Category!.Name,
                        CategorySlug = p.Category.Slug
                    })
                    .ToListAsync();

                return Ok(new
                {
                    success = true,
                    data = posts,
                    pagination = new
                    {
                        page,
                        limit,
                        total,
                        pages = (int)Math.Ceiling((double)total / limit)
                    }
                });
            }
            catch (Exception ex)
            {
                return StatusCode(500, new
                {
                    success = false,
                    message = "Error interno del servidor"
                });
            }
        }

        // GET: api/posts/{slug}/related
        [HttpGet("{slug}/related")]
        public async Task<ActionResult<object>> GetRelatedPosts(
            string slug, 
            [FromQuery] int limit = 3)
        {
            try
            {
                var originalPost = await _context.Posts
                    .Where(p => p.Slug == slug)
                    .Select(p => new { p.CategoryId })
                    .FirstOrDefaultAsync();

                if (originalPost == null)
                {
                    return NotFound(new
                    {
                        success = false,
                        message = "Post no encontrado"
                    });
                }

                var relatedPosts = await _context.Posts
                    .Include(p => p.Category)
                    .Where(p => p.Status == "published" &&
                               p.Slug != slug &&
                               p.CategoryId == originalPost.CategoryId)
                    .OrderByDescending(p => p.PublishedAt)
                    .Take(limit)
                    .Select(p => new
                    {
                        p.Id,
                        p.Title,
                        p.Slug,
                        p.Excerpt,
                        p.CreatedAt,
                        CategoryName = p.Category!.Name,
                        CategorySlug = p.Category.Slug
                    })
                    .ToListAsync();

                return Ok(new
                {
                    success = true,
                    data = relatedPosts
                });
            }
            catch (Exception ex)
            {
                return StatusCode(500, new
                {
                    success = false,
                    message = "Error interno del servidor"
                });
            }
        }
    }
} 